src/track-meta-data.cc

```diff
@@ -33,95 +34,96 @@
-      gchar* value = nullptr;
-      if (!gst_tag_list_get_string(tag_list, tag_name, &value)) return;
-
-      Context* context = (Context*) ctx;
-      std::string &to_update = (*context->fields)[xml_name];
-      if (to_update != value) {
-        context->modified = true;
-        to_update = value;
-      }
-      g_free(value);
-    }, &ctx);
-
-  return ctx.modified;
-}
+  @param  xml_document pugi::xml_document to add root item to
+  @retval none
+*/
+void TrackMetadata::CreateXmlRoot(XMLDoc& xml_document) const {
+  XMLElement root = xml_document.AddElement("DIDL-Lite");
 
-bool TrackMetadata::ParseXML(const std::string &xml) {
-  const auto doc = XMLDoc::Parse(xml);
-  if (!doc) return false;
+  root.SetAttribute("xmlns", "urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/");
+  root.SetAttribute("xmlns:dc", "http://purl.org/dc/elements/1.1/");
+  root.SetAttribute("xmlns:upnp", "urn:schemas-upnp-org:metadata-1-0/upnp/");
 
-  const auto items = doc->findElement("DIDL-Lite").findElement("item");
-  for (XMLElement elem : items.children()) {
-    fields_[elem.name()] = elem.value();
-  }
-  return items.exists();
+  // An "item" element must have
+  // dc:title element first
+  // upnp:class element
+  // id attribute
+  // parentId attribute
+  // restricted attribute
+  XMLElement item = root.AddElement("item");
+  item.SetAttribute("id", "");
+  item.SetAttribute("parentID", "0");
+  item.SetAttribute("restricted", "false");
 }
 
-/*static*/ std::string TrackMetadata::DefaultCreateNewId() {
-  // Generating a unique ID in case the players cache the content by
-  // the item-ID. Right now this is experimental and not known to make
-  // any difference - it seems that players just don't display changes
-  // in the input stream. Grmbl.
-  static unsigned int xml_id = 42;
-  char unique_id[4 + 8 + 1];
-  snprintf(unique_id, sizeof(unique_id), "gmr-%08x", xml_id++);
-  return unique_id;
-}
+/**
+  @brief  Format metadata to XML (DIDL-Lite) format. Modifies existing XML
+          if passed as argument
+
+  @param  xml Original XML string to modify
+  @retval std::string Metadata as XML (DIDL-Lite)
+*/
+std::string TrackMetadata::ToXml(const std::string& xml) const {
+  // Parse existing document
+  auto xml_document = XMLDoc::Parse(xml);
 
-std::string TrackMetadata::generateDIDL(const std::string &id) const {
-  XMLDoc doc;
-  auto item = doc.AddElement("DIDL-Lite",
-                             "urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/")
-    .SetAttribute("xmlns:dc", "http://purl.org/dc/elements/1.1/")
-    .SetAttribute("xmlns:upnp", "urn:schemas-upnp-org:metadata-1-0/upnp/")
-    .AddElement("item").SetAttribute("id", id);
-  for (auto pair : fields_) {
-    item.AddElement(pair.first).SetValue(pair.second);
+  XMLElement root;
+  XMLElement item;
+
+  // Attempt to find root and item element from original XML
+  if (xml_document != nullptr)
+  {
+    root = xml_document->findElement("DIDL-Lite");
+    item = root.findElement("item");
```

Comment:
This is unnecessary complicated. Have a look at the original implementaiton where I just chain these:
`items = doc->findElement("DIDL-Lite").findElement("item")`

There are other places that look more complicated in your implementation, I suppose this is a carry-over from the re-basing. Try to only change the things that are needed for the additional date and tracknumber support.